File: back/services/src/Api/Models/Filter.cs
namespace Api.Models
{
    public class Filter
    {
        public string Query;
        public SkillsFilter Skills { get; set; } = new SkillsFilter();
        public TablewaresFilter Tablewares { get; set; } = new TablewaresFilter();
        public int? Personas { get; set; }
        public TimesFilter Times { get; set; } = new TimesFilter();
        public TypesFilter Types { get; set; } = new TypesFilter();
    }

    public class SkillsFilter
    {
        public bool IsEnabled => IsEasy || IsMedium || IsHard;
        public bool IsEasy { get; set; } = false;
        public bool IsMedium { get; set; } = false;
        public bool IsHard { get; set; } = false;
    }

    public class TablewaresFilter
    {
        public bool IsEnabled => HasPan || HasOven || HasFryingPan;
        public bool HasPan { get; set; } = false;
        public bool HasOven { get; set; } = false;
        public bool HasFryingPan { get; set; } = false;
    }

    public class TimesFilter
    {
        public bool IsEnabled => Is10Min || IsFrom10To30Min || IsMoreThan30Min;
        public bool Is10Min { get; set; } = false;
        public bool IsFrom10To30Min { get; set; } = false;
        public bool IsMoreThan30Min { get; set; } = false;
    }

    public class TypesFilter
    {
        public bool IsEnabled => IsSweet || IsMeat || IsVegan;
        public bool IsSweet { get; set; } = false;
        public bool IsMeat { get; set; } = false;
        public bool IsVegan { get; set; } = false;
    }
}

File: app/lib/models/filter.dart
class Filter {
  String query;
  SkillsFilter skills = SkillsFilter();
  TablewaresFilter tablewares = TablewaresFilter();
  int personas = 0;
  TimesFilter times = TimesFilter();
  TypesFilter types = TypesFilter();

  Map<String, dynamic> toMap() {
    return {
      'query': query,
      'personas': personas,
      'skills': skills.toMap(),
      'tablewares': tablewares.toMap(),
      'times': times.toMap(),
      'types': types.toMap()
    };
  }
}

class SkillsFilter {
  bool get isEnabled => isEasy || isMedium || isHard;
  bool isEasy = false;
  bool isMedium = false;
  bool isHard = false;

  Map<String, dynamic> toMap() {
    return {
      'isEasy': isEasy,
      'isMedium': isMedium,
      'isHard': isHard,
    };
  }
}

class TablewaresFilter {
  bool get isEnabled => hasPan || hasOven || hasFryingPan;
  bool hasPan = false;
  bool hasOven = false;
  bool hasFryingPan = false;

  Map<String, dynamic> toMap() {
    return {
      'hasPan': hasPan,
      'hasOven': hasOven,
      'hasFryingPan': hasFryingPan,
    };
  }
}

class TimesFilter {
  bool get isEnabled => is10Min || isFrom10To30Min || isMoreThan30Min;
  bool is10Min = false;
  bool isFrom10To30Min = false;
  bool isMoreThan30Min = false;

  Map<String, dynamic> toMap() {
    return {
      'is10Min': is10Min,
      'isFrom10To30Min': isFrom10To30Min,
      'isMoreThan30Min': isMoreThan30Min,
    };
  }
}

class TypesFilter {
  bool get isEnabled => isSweet || isMeat || isVegan;
  bool isSweet = false;
  bool isMeat = false;
  bool isVegan = false;

  Map<String, dynamic> toMap() {
    return {
      'isSweet': isSweet,
      'isMeat': isMeat,
      'isVegan': isVegan,
    };
  }
}
